File: matlab/tests/unit/test_ENoB_bitSweep.m
%% test_ENoB_bitSweep.m - Unit test for ENoB_bitSweep function
% Output: test_output/<dataset>/test_ENoB_bitSweep/
%   - ENoB_sweep_matlab.png
%   - ENoB_sweep_data_matlab.csv

close all; clc; clear;

%% Configuration
inputDir = "dataset";
outputDir = "test_output";
filesList = autoSearchFiles({}, inputDir, 'dout_*.csv');

if ~isfolder(outputDir)
    mkdir(outputDir);
end

%% Test Loop
fprintf('=== test_ENoB_bitSweep.m ===\n');
fprintf('[Testing] %d datasets...\n\n', length(filesList));

for k = 1:length(filesList)
    currentFilename = filesList{k};
    dataFilePath = fullfile(inputDir, currentFilename);

    if ~isfile(dataFilePath)
        fprintf('[%d/%d] %s - NOT FOUND\n\n', k, length(filesList), currentFilename);
        continue;
    end
    fprintf('[%d/%d] [Processing] %s\n', k, length(filesList), currentFilename);

    [~, datasetName, ~] = fileparts(currentFilename);
    subFolder = fullfile(outputDir, datasetName, 'test_ENoB_bitSweep');
    if ~isfolder(subFolder)
        mkdir(subFolder);
    end

    %% Run ENoB_bitSweep
    figure('Position', [100, 100, 800, 600], 'Visible', 'off');
    [ENoB_sweep, nBits_vec] = ENoB_bitSweep(readmatrix(dataFilePath), ...
        'freq', 0, 'order', 5, 'harmonic', 5, 'OSR', 1, 'winType', @hamming);

    title(replace(datasetName, '_', '\_'));

    % Save results
    saveas(gcf, fullfile(subFolder, 'ENoB_sweep_matlab.png'));
    fprintf('  [Saved] ENoB_sweep_matlab.png\n');
    % close(gcf);

    writetable(table(nBits_vec', ENoB_sweep', 'VariableNames', {'nBits', 'ENoB'}), ...
        fullfile(subFolder, 'ENoB_sweep_data_matlab.csv'));
    fprintf('  [Saved] ENoB_sweep_data_matlab.csv\n');

    % Summary
    maxENoB = max(ENoB_sweep(~isnan(ENoB_sweep)));
    fprintf('  [Results] Max ENoB = %.2f bits (using %d bits)\n\n', ...
        maxENoB, find(ENoB_sweep == maxENoB, 1));
end

fprintf('[test_ENoB_bitSweep COMPLETE]\n');

File: python/tests/unit/test_enob_bit_sweep.py
import matplotlib.pyplot as plt

from adctoolbox.dout.enob_bit_sweep import enob_bit_sweep
from tests._utils import save_variable, save_fig
from tests.unit._runner import run_unit_test_batch

plt.rcParams['font.size'] = 14
plt.rcParams['axes.grid'] = True

def _process_enob_bit_sweep(raw_data, sub_folder, dataset_name):
    """
    Callback function to process a single file:
    1. Run ENOB bit sweep analysis
    2. Save ENOB sweep and nBits vectors
    3. Save plot
    """
    # Create figure and run enob_bit_sweep
    fig = plt.figure(figsize=(10, 7.5))
    enob_sweep, n_bits_vec = enob_bit_sweep(
        raw_data, freq=0, order=5, harmonic=5, osr=1, win_type=4, plot=True)

    # Save figure
    save_fig(sub_folder, 'ENoB_bitSweep.png', dpi=150)

    # Save variables
    save_variable(sub_folder, enob_sweep, 'ENoB_sweep')
    save_variable(sub_folder, n_bits_vec, 'nBits_vec')

def test_enob_bit_sweep(project_root):
    """
    Batch runner for ENOB bit sweep analysis.
    """
    run_unit_test_batch(
        project_root=project_root,
        input_subpath="dataset/dout",
        test_module_name="test_enob_bit_sweep",
        file_pattern="dout_*.csv",
        output_subpath="test_output",
        process_callback=_process_enob_bit_sweep,
        flatten=False  # Digital output data is 2D (N samples x M bits)
    )
